libdrgn/iterators.c

```diff
@@ -0,0 +1,472 @@
+		err = drgn_object_member_dereference(&iter->root, root,
+						     "rnode");
+		if (err)
+			goto out;
+		err = drgn_program_find_type(prog, "void *", NULL,
+					     &iter->node_type);
+		if (err)
+			goto out;
+		err = drgn_object_cast(&iter->root, iter->node_type,
+				       &iter->root);
+		if (err)
+			goto out;
+		err = drgn_program_find_type(prog, "struct radix_tree_node *",
+					     NULL, &iter->node_type);
+		if (err)
+			goto out;
+		iter->RADIX_TREE_INTERNAL_NODE = 1;
+	} else {
+		goto out;
+	}
+
+	struct drgn_type_member *member;
+	uint64_t member_bit_offset;
+	err = drgn_type_find_member(drgn_type_type(iter->node_type.type).type,
+				    "slots", &member, &member_bit_offset);
+	if (err)
+		goto out;
+	struct drgn_qualified_type member_type;
+	err = drgn_member_type(member, &member_type, NULL);
+	if (err)
+		goto out;
+	if (drgn_type_kind(member_type.type) != DRGN_TYPE_ARRAY) {
+		err = drgn_error_create(
+			DRGN_ERROR_TYPE,
+			"struct radix_tree_node slots member is not an array");
+		goto out;
+	}
+	iter->RADIX_TREE_MAP_MASK = drgn_type_length(member_type.type) - 1;
+
+out:
+	if (err) {
+		drgn_object_deinit(&iter->entry.node);
+		frames_deinit(&iter->frames);
+	}
+	drgn_object_deinit(&tmp);
+	return err;
+}
+
+void radix_tree_iter_deinit(struct radix_tree_iter *iter)
+{
+	drgn_object_deinit(&iter->root);
+	drgn_object_deinit(&iter->entry.node);
+	while (iter->frames.size) {
+		drgn_object_deinit(&frames_pop(&iter->frames)->slots);
+	}
+	frames_deinit(&iter->frames);
+}
+
+struct drgn_error *radix_tree_iter_next(struct radix_tree_iter *iter,
+					struct radix_tree_iter_entry **ret)
+{
+	bool entry_populated = false;
+	bool node_is_truthy = false;
+	struct drgn_error *err = NULL;
+	struct drgn_object node;
+	drgn_object_init(&node, iter->prog);
+	if (!iter->started) {
+		iter->started = true;
+		err = radix_tree_iter_handle_node(iter, &iter->root, 0,
+						  &entry_populated);
+		if (!err && entry_populated)
+			err = drgn_object_bool(&iter->entry.node,
+					       &node_is_truthy);
+	}
+
+	while (!err && !entry_populated && iter->frames.size) {
+		struct radix_tree_iter_frame *frame =
+			&iter->frames.data[iter->frames.size - 1];
+		if (frame->loop_index <= iter->RADIX_TREE_MAP_MASK) {
+			err = drgn_object_subscript(&node, &frame->slots,
+						    frame->loop_index);
+			if (!err)
+				err = radix_tree_iter_handle_node(
+					iter, &node,
+					frame->index + (frame->loop_index++
+							<< frame->shift),
+					&entry_populated);
+		} else {
+			drgn_object_deinit(&frame->slots);
+			frames_pop(&iter->frames);
+		}
+	}
+	*ret = err || !entry_populated ? NULL : &iter->entry;
+	drgn_object_deinit(&node);
+	return err;
+}
+
+struct drgn_error *idr_iter_init(struct idr_iter *iter,
+				 const struct drgn_object *idr)
+{
+	struct drgn_object idr_rt;
+	struct drgn_object idr_base;
+	struct drgn_error *err;
+	drgn_object_init(&idr_rt, drgn_object_program(idr));
+	drgn_object_init(&idr_base, drgn_object_program(idr));
```

Comment:
FWIW, it took me _much_ longer to understand `linux_helper_radix_tree_lookup` when I was referencing it to aid with my translation of `radix_tree_for_each` because it uses two `drgn_object`s (`node` and `tmp`) for _many_ different purposes, which was extremely confusing and was only made clear by having the Python implementation of `radix_tree_for_each` to reference because it _did_ use unique variable names.

I don't think the very minor decrease in memory usage here is worth the harm to readability (comments only go so far).